src/spartan_handler_fsm.erl

```diff
@@ -244,54 +217,55 @@ reply_fail(_State1 = #state{dns_message = DNSMessage, from = {FromModule, FromKe
 
 %% @private
 resolve(Parent, Upstream, State) ->
-    try do_resolve(Parent, Upstream, State) of
-        _ ->
-            ok
+    try
+        do_resolve(Parent, Upstream, State)
     catch
-        exit:Exception ->
-            lager:warning("Resolver (~p) Process exited: ~p stacktrace ~p", [Upstream, Exception, erlang:get_stacktrace()]),
-            %% Reraise the exception
-            exit(Exception);
-        Reason ->
-            lager:warning("Resolver (~p) Process exited: ~p stacktrace ~p", [Upstream, Reason, erlang:get_stacktrace()]),
-            exit(unknown_error)
+        Class:Reason ->
+            StackTrace = erlang:get_stacktrace(),
+            lager:warning("Resolver (~p) Process exited: ~p stacktrace ~p",
+                [Upstream, Reason, erlang:get_stacktrace()]),
+            erlang:raise(Class, Reason, StackTrace)
     end.
 
 do_resolve(Parent, Upstream = {UpstreamIP, UpstreamPort}, #state{data = Data, from = {spartan_udp_server, _}}) ->
-    lager:debug("Sending query to Upstream: ~p", [Upstream]),
-    {ok, _} = timer:kill_after(?TIMEOUT),
     {ok, Socket} = gen_udp:open(0, [{reuseaddr, true}, {active, once}, binary]),
-    link(Socket),
     gen_udp:send(Socket, UpstreamIP, UpstreamPort, Data),
-    %% Should put a timeout here given we're linked to our parents?
-    receive
-        {udp, Socket, UpstreamIP, UpstreamPort, ReplyData} ->
-            lager:debug("Received Reply"),
-            gen_fsm:send_event(Parent, {upstream_reply, Upstream, ReplyData});
-        Else ->
-            lager:debug("Received else: ~p, while upstream: ~p", [Else, Upstream])
+    MonRef = erlang:monitor(process, Parent),
+    try
+        receive
+            {'DOWN', MonRef, process, _Pid, Reason} ->
+                exit(Reason);
+            {udp, Socket, UpstreamIP, UpstreamPort, ReplyData} ->
+                gen_fsm:send_event(Parent, {upstream_reply, Upstream, ReplyData})
         after ?TIMEOUT ->
-            lager:debug("Timed out waiting for upstream: ~p", [Upstream])
```

Comment:
The drawback of recon trace is that you have to know what you're looking for, whereas a message like this is a good indicator of what may be going wrong.

I guess with the metrics it's ok to take this out though